File: commands.go
package main

import (
	"database/sql"
	"fmt"

	"github.com/disgoorg/disgo/bot"
	"github.com/disgoorg/disgo/discord"
	"github.com/disgoorg/disgo/events"
	"github.com/disgoorg/snowflake/v2"
)

var commands = []discord.ApplicationCommandCreate{
	discord.SlashCommandCreate{
		Name:        "about",
		Description: "Information about the bot",
	},
	discord.SlashCommandCreate{
		Name:        "current-count",
		Description: "Get the current count",
	},
	discord.SlashCommandCreate{
		Name:        "rule",
		Description: "Manage counting rules",
		Options: []discord.ApplicationCommandOption{
			discord.ApplicationCommandOptionSubCommand{
				Name:        "create",
				Description: "Create a new rule",
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionString{
						Name:        "action",
						Description: "Action to perform",
						Required:    true,
						Choices: []discord.ApplicationCommandOptionChoiceString{
							{Name: "Pin Message", Value: "pin"},
							{Name: "Send DM", Value: "dm"},
							{Name: "Send Message", Value: "msg"},
						},
					},
					discord.ApplicationCommandOptionString{
						Name:        "type",
						Description: "Condition type",
						Required:    true,
						Choices: []discord.ApplicationCommandOptionChoiceString{
							{Name: "Equals", Value: "equals"},
							{Name: "Multiple Of", Value: "multiple_of"},
						},
					},
					discord.ApplicationCommandOptionInt{
						Name:        "value",
						Description: "Value for condition",
						Required:    true,
					},
					discord.ApplicationCommandOptionString{
						Name:        "content",
						Description: "Content to send (use {{count}} for count) (ignored for pin)",
						Required:    false,
					},
				},
			},
			discord.ApplicationCommandOptionSubCommand{
				Name:        "delete",
				Description: "Delete a specific rule",
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionString{
						Name:        "id",
						Description: "Rule ID",
						Required:    true,
					},
				},
			},
			discord.ApplicationCommandOptionSubCommand{
				Name:        "view",
				Description: "View a specific rule",
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionString{
						Name:        "id",
						Description: "Rule ID",
						Required:    true,
					},
				},
			},
			discord.ApplicationCommandOptionSubCommand{
				Name:        "list",
				Description: "List all rules",
			},
		},
	},
	discord.SlashCommandCreate{
		Name:        "settings",
		Description: "Configure tally settings",
		Options: []discord.ApplicationCommandOption{
			discord.ApplicationCommandOptionSubCommand{
				Name:        "current-count",
				Description: "Set the current count",
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionInt{
						Name:        "number",
						Description: "The count to set",
						Required:    true,
					},
					discord.ApplicationCommandOptionString{
						Name:        "purge",
						Description: "Purge messages down to this count?",
						Required:    false,
						Choices: []discord.ApplicationCommandOptionChoiceString{
							{Name: "Yes", Value: "yes"},
							{Name: "No", Value: "no"},
						},
					},
				},
			},
			discord.ApplicationCommandOptionSubCommand{
				Name:        "channel",
				Description: "Set the counting channel",
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionChannel{
						Name:        "channel",
						Description: "The channel to count in",
						Required:    true,
						ChannelTypes: []discord.ChannelType{
							discord.ChannelTypeGuildText,
						},
					},
				},
			},
			discord.ApplicationCommandOptionSubCommand{
				Name:        "take-turns",
				Description: "Toggle consecutive counting",
				Options: []discord.ApplicationCommandOption{
					discord.ApplicationCommandOptionString{
						Name:        "state",
						Description: "State (on forces taking turns, off allows consecutive counting)",
						Required:    true,
						Choices: []discord.ApplicationCommandOptionChoiceString{
							{Name: "On", Value: "on"},
							{Name: "Off", Value: "off"},
						},
					},
				},
			},
			discord.ApplicationCommandOptionSubCommand{
				Name:        "view",
				Description: "View current settings",
			},
		},
	},
}

func registerCommands(client *bot.Client) {
	_, err := client.Rest.SetGlobalCommands(client.ApplicationID, commands)
	if err != nil {
		fmt.Println("error registering commands: ", err)
	}
}

func commandHandler(db *sql.DB) func(event *events.ApplicationCommandInteractionCreate) {
	return func(event *events.ApplicationCommandInteractionCreate) {
		data := event.SlashCommandInteractionData()
		switch data.CommandName() {
		case "about":
			content := fmt.Sprintf("Tally counting bot rewritten in Go.\n\n**Help:** Start counting by typing numbers in the designated channel.\n**Invite:** https://discord.com/api/oauth2/authorize?client_id=%s&permissions=8&scope=bot%%20applications.commands\n**Ping:** %dms", event.Client().ApplicationID.String(), event.Client().Gateway.Latency().Milliseconds())
			event.CreateMessage(discord.MessageCreate{
				Content: content,
				Flags:   discord.MessageFlagEphemeral,
			})
		case "current-count":
			if event.GuildID() == nil {
				return
			}
			entry, _ := getCountEntry(db, event.GuildID().String())
			event.CreateMessage(discord.MessageCreate{
				Content: fmt.Sprintf("The current count is %d", entry.Count),
				Flags:   discord.MessageFlagEphemeral,
			})
		case "rule":
			if event.GuildID() == nil {
				return
			}
			sub := *data.SubCommandName
			switch sub {
			case "create":
				action := data.String("action")
				typ := data.String("type")
				val := data.Int("value")
				content := data.String("content")
				if content == "" && action != "pin" {
					content = "Triggered rule at count {{count}}!"
				}
				id := fmt.Sprintf("%x", event.ID())
				r := GuildRule{
					ID:       id,
					Guild:    event.GuildID().String(),
					Trigger:  "count",
					Type:     typ,
					Value:    val,
					Action:   action,
					ActionV1: content,
				}
				_ = addGuildRule(db, r)
				event.CreateMessage(discord.MessageCreate{
					Content: fmt.Sprintf("Created rule `%s`:\nAction: %s\nCondition: %s %d", id, action, typ, val),
					Flags:   discord.MessageFlagEphemeral,
				})
			case "delete":
				id := data.String("id")
				_ = deleteGuildRule(db, id, event.GuildID().String())
				event.CreateMessage(discord.MessageCreate{
					Content: fmt.Sprintf("Deleted rule `%s`", id),
					Flags:   discord.MessageFlagEphemeral,
				})
			case "view":
				id := data.String("id")
				r, err := getGuildRule(db, id, event.GuildID().String())
				if err != nil {
					event.CreateMessage(discord.MessageCreate{
						Content: "Rule not found.",
						Flags:   discord.MessageFlagEphemeral,
					})
					return
				}
				event.CreateMessage(discord.MessageCreate{
					Content: fmt.Sprintf("Rule `%s`:\nAction: %s\nType: %s\nValue: %d\nContent: %s", r.ID, r.Action, r.Type, r.Value, r.ActionV1),
					Flags:   discord.MessageFlagEphemeral,
				})
			case "list":
				rules, _ := getGuildRules(db, event.GuildID().String())
				if len(rules) == 0 {
					event.CreateMessage(discord.MessageCreate{
						Content: "No rules configured.",
						Flags:   discord.MessageFlagEphemeral,
					})
					return
				}
				out := "Rules:\n"
				for _, r := range rules {
					out += fmt.Sprintf("- `%s`: %s on %s %d\n", r.ID, r.Action, r.Type, r.Value)
				}
				event.CreateMessage(discord.MessageCreate{
					Content: out,
					Flags:   discord.MessageFlagEphemeral,
				})
			}
		case "settings":
			if event.GuildID() == nil {
				return
			}
			sub := *data.SubCommandName
			switch sub {
			case "current-count":
				_ = event.DeferCreateMessage(true)
				
				num := data.Int("number")

				purgeOpt, ok := data.OptString("purge")
				purge := ok && purgeOpt == "yes"

				if purge {
					config, _ := getGuildConfig(db, event.GuildID().String())
					webhookID, _ := snowflake.Parse(config.WebhookID)

					msgs, _ := getMessagesToPurge(db, event.GuildID().String(), num)
					for _, msgIDStr := range msgs {
						if msgIDStr == "" {
							continue
						}
						msgID, _ := snowflake.Parse(msgIDStr)
						_ = event.Client().Rest.DeleteWebhookMessage(webhookID, config.WebhookToken, msgID, 0)
					}
					_ = deletePurgedMessages(db, event.GuildID().String(), num)
				}

				entry, _ := getCountEntry(db, event.GuildID().String())
				entry.Count = num
				entry.LastCounter = ""
				setCountEntry(db, entry)

				msg := fmt.Sprintf("Count set to %d.", num)
				if purge {
					msg += " Purged newer messages."
				}
				
				_, _ = event.Client().Rest.UpdateInteractionResponse(event.ApplicationID(), event.Token(), discord.MessageUpdate{
					Content: &msg,
				})
			case "channel":
				channelID := data.Snowflake("channel")
				config, _ := getGuildConfig(db, event.GuildID().String())

				webhook, err := createCountingWebhook(event.Client(), channelID)
				if err != nil {
					event.CreateMessage(discord.MessageCreate{
						Content: "Failed to create webhook in channel.",
						Flags:   discord.MessageFlagEphemeral,
					})
					return
				}

				config.Active = true
				config.Channel = channelID.String()
				config.WebhookID = webhook.ID().String()
				config.WebhookToken = webhook.(*discord.IncomingWebhook).Token
				setGuildConfig(db, config)
				event.CreateMessage(discord.MessageCreate{
					Content: fmt.Sprintf("Counting channel set to <#%s>", channelID.String()),
					Flags:   discord.MessageFlagEphemeral,
				})
			case "take-turns":
				state := data.String("state")
				config, _ := getGuildConfig(db, event.GuildID().String())
				if state == "on" {
					config.AllowDoublePost = false
				} else {
					config.AllowDoublePost = true
				}
				setGuildConfig(db, config)
				event.CreateMessage(discord.MessageCreate{
					Content: fmt.Sprintf("Take turns setting is now %s.", state),
					Flags:   discord.MessageFlagEphemeral,
				})
			case "view":
				config, _ := getGuildConfig(db, event.GuildID().String())

				channelStr := "None"
				if config.Channel != "" {
					channelStr = fmt.Sprintf("<#%s>", config.Channel)
				}

				takeTurns := "On"
				if config.AllowDoublePost {
					takeTurns = "Off"
				}

				out := fmt.Sprintf("**Tally Settings:**\n\n**Counting Channel:** %s\n**Take Turns:** %s", channelStr, takeTurns)

				event.CreateMessage(discord.MessageCreate{
					Content: out,
					Flags:   discord.MessageFlagEphemeral,
				})
			}
		}
	}
}

File: src/commands.rs
use twilight_model::{
    application::command::{Command, CommandType},
    channel::ChannelType,
    application::interaction::{Interaction, InteractionData},
    application::interaction::application_command::CommandOptionValue,
    http::interaction::{InteractionResponse, InteractionResponseData, InteractionResponseType},
    id::marker::WebhookMarker,
    id::Id,
};
use twilight_util::builder::command::{
    ChannelBuilder, CommandBuilder, IntegerBuilder, StringBuilder, SubCommandBuilder,
};
use std::sync::{Arc, Mutex};
use rusqlite::Connection;
use twilight_http::Client as HttpClient;

pub fn get_commands() -> Vec<Command> {
    vec![
        CommandBuilder::new("about", "Information about the bot", CommandType::ChatInput).build(),
        CommandBuilder::new("current-count", "Get the current count", CommandType::ChatInput).build(),
        CommandBuilder::new("rule", "Manage counting rules", CommandType::ChatInput)
            .option(
                SubCommandBuilder::new("create", "Create a new rule")
                    .option(
                        StringBuilder::new("action", "Action to perform")
                            .required(true)
                            .choices(vec![
                                ("Pin Message".to_string(), "pin".to_string()),
                                ("Send DM".to_string(), "dm".to_string()),
                                ("Send Message".to_string(), "msg".to_string()),
                            ])
                    )
                    .option(
                        StringBuilder::new("type", "Condition type")
                            .required(true)
                            .choices(vec![
                                ("Equals".to_string(), "equals".to_string()),
                                ("Multiple Of".to_string(), "multiple_of".to_string()),
                            ])
                    )
                    .option(IntegerBuilder::new("value", "Value for condition").required(true))
                    .option(StringBuilder::new("content", "Content to send (use {{count}} for count) (ignored for pin)").required(false))
            )
            .option(
                SubCommandBuilder::new("delete", "Delete a specific rule")
                    .option(StringBuilder::new("id", "Rule ID").required(true))
            )
            .option(
                SubCommandBuilder::new("view", "View a specific rule")
                    .option(StringBuilder::new("id", "Rule ID").required(true))
            )
            .option(SubCommandBuilder::new("list", "List all rules"))
            .build(),
        CommandBuilder::new("settings", "Configure tally settings", CommandType::ChatInput)
            .option(
                SubCommandBuilder::new("current-count", "Set the current count")
                    .option(IntegerBuilder::new("number", "The count to set").required(true))
                    .option(
                        StringBuilder::new("purge", "Purge messages down to this count?")
                            .required(false)
                            .choices(vec![
                                ("Yes".to_string(), "yes".to_string()),
                                ("No".to_string(), "no".to_string()),
                            ])
                    )
            )
            .option(
                SubCommandBuilder::new("channel", "Set the counting channel")
                    .option(
                        ChannelBuilder::new("channel", "The channel to count in")
                            .required(true)
                            .channel_types(vec![ChannelType::GuildText])
                    )
            )
            .option(
                SubCommandBuilder::new("take-turns", "Toggle consecutive counting")
                    .option(
                        StringBuilder::new("state", "State (on forces taking turns, off allows consecutive counting)")
                            .required(true)
                            .choices(vec![
                                ("On".to_string(), "on".to_string()),
                                ("Off".to_string(), "off".to_string()),
                            ])
                    )
            )
            .option(SubCommandBuilder::new("view", "View current settings"))
            .build(),
    ]
}

pub async fn handle_interaction(
    db: Arc<Mutex<Connection>>,
    http: Arc<HttpClient>,
    interaction: Box<Interaction>,
) -> Result<(), Box<dyn std::error::Error + Send + Sync>> {
    let app_id = interaction.application_id;
    let interaction_id = interaction.id;
    let interaction_token = interaction.token.clone();
    
    if let Some(InteractionData::ApplicationCommand(data)) = &interaction.data {
        let response_data = match data.name.as_str() {
            "about" => {
                Some(InteractionResponseData {
                    content: Some(format!("Tally counting bot rewritten in Rust.")),
                    flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                    ..Default::default()
                })
            }
            "current-count" => {
                if let Some(guild_id) = interaction.guild_id {
                    let entry = {
                        let conn = db.lock().unwrap();
                        crate::database::get_count_entry(&conn, &guild_id.get().to_string()).unwrap()
                    };
                    Some(InteractionResponseData {
                        content: Some(format!("The current count is {}", entry.count)),
                        flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                        ..Default::default()
                    })
                } else {
                    None
                }
            }
            "rule" => {
                if let Some(guild_id) = interaction.guild_id {
                    if let Some(opt) = data.options.first() {
                        match &opt.value {
                            CommandOptionValue::SubCommand(opts) => {
                                match opt.name.as_str() {
                                    "create" => {
                                        let mut action = String::new();
                                        let mut typ = String::new();
                                        let mut val = 0;
                                        let mut content = String::new();
                                        for o in opts {
                                            match o.name.as_str() {
                                                "action" => if let CommandOptionValue::String(s) = &o.value { action = s.clone(); },
                                                "type" => if let CommandOptionValue::String(s) = &o.value { typ = s.clone(); },
                                                "value" => if let CommandOptionValue::Integer(i) = &o.value { val = *i; },
                                                "content" => if let CommandOptionValue::String(s) = &o.value { content = s.clone(); },
                                                _ => {}
                                            }
                                        }
                                        if content.is_empty() && action != "pin" {
                                            content = "Triggered rule at count {{count}}!".to_string();
                                        }
                                        let id = format!("{:x}", interaction.id.get());
                                        let r = crate::models::GuildRule {
                                            id: id.clone(),
                                            guild: guild_id.get().to_string(),
                                            trigger: "count".to_string(),
                                            rule_type: typ.clone(),
                                            value: val,
                                            action: action.clone(),
                                            action_v1: content,
                                        };
                                        {
                                            let conn = db.lock().unwrap();
                                            let _ = crate::database::add_guild_rule(&conn, &r);
                                        }
                                        Some(InteractionResponseData {
                                            content: Some(format!("Created rule `{}`:\nAction: {}\nCondition: {} {}", id, action, typ, val)),
                                            flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                            ..Default::default()
                                        })
                                    }
                                    "delete" => {
                                        let mut id = String::new();
                                        for o in opts {
                                            if o.name == "id" {
                                                if let CommandOptionValue::String(s) = &o.value { id = s.clone(); }
                                            }
                                        }
                                        {
                                            let conn = db.lock().unwrap();
                                            let _ = crate::database::delete_guild_rule(&conn, &id, &guild_id.get().to_string());
                                        }
                                        Some(InteractionResponseData {
                                            content: Some(format!("Deleted rule `{}`", id)),
                                            flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                            ..Default::default()
                                        })
                                    }
                                    "view" => {
                                        let mut id = String::new();
                                        for o in opts {
                                            if o.name == "id" {
                                                if let CommandOptionValue::String(s) = &o.value { id = s.clone(); }
                                            }
                                        }
                                        let rule = {
                                            let conn = db.lock().unwrap();
                                            crate::database::get_guild_rule(&conn, &id, &guild_id.get().to_string())
                                        };
                                        if let Ok(r) = rule {
                                            Some(InteractionResponseData {
                                                content: Some(format!("Rule `{}`:\nAction: {}\nType: {}\nValue: {}\nContent: {}", r.id, r.action, r.rule_type, r.value, r.action_v1)),
                                                flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                ..Default::default()
                                            })
                                        } else {
                                            Some(InteractionResponseData {
                                                content: Some("Rule not found.".to_string()),
                                                flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                ..Default::default()
                                            })
                                        }
                                    }
                                    "list" => {
                                        let rules = {
                                            let conn = db.lock().unwrap();
                                            crate::database::get_guild_rules(&conn, &guild_id.get().to_string()).unwrap_or_default()
                                        };
                                        if rules.is_empty() {
                                            Some(InteractionResponseData {
                                                content: Some("No rules configured.".to_string()),
                                                flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                ..Default::default()
                                            })
                                        } else {
                                            let mut out = "Rules:\n".to_string();
                                            for r in rules {
                                                out.push_str(&format!("- `{}`: {} on {} {}\n", r.id, r.action, r.rule_type, r.value));
                                            }
                                            Some(InteractionResponseData {
                                                content: Some(out),
                                                flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                ..Default::default()
                                            })
                                        }
                                    }
                                    _ => None
                                }
                            }
                            _ => None
                        }
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
            "settings" => {
                if let Some(guild_id) = interaction.guild_id {
                    if let Some(opt) = data.options.first() {
                        match &opt.value {
                            CommandOptionValue::SubCommand(opts) => {
                                match opt.name.as_str() {
                                    "current-count" => {
                                        let mut num = 0;
                                        let mut purge = false;
                                        for o in opts {
                                            match o.name.as_str() {
                                                "number" => if let CommandOptionValue::Integer(i) = &o.value { num = *i; },
                                                "purge" => if let CommandOptionValue::String(s) = &o.value { purge = s == "yes"; },
                                                _ => {}
                                            }
                                        }
                                        if purge {
                                            let config = {
                                                let conn = db.lock().unwrap();
                                                crate::database::get_guild_config(&conn, &guild_id.get().to_string()).unwrap()
                                            };
                                            let webhook_id_parsed: u64 = config.webhook_id.parse().unwrap_or(0);
                                            if webhook_id_parsed != 0 {
                                                let webhook_id = Id::<WebhookMarker>::new(webhook_id_parsed);
                                                let msgs = {
                                                    let conn = db.lock().unwrap();
                                                    crate::database::get_messages_to_purge(&conn, &guild_id.get().to_string(), num).unwrap_or_default()
                                                };
                                                for msg_id_str in msgs {
                                                    if let Ok(m_id) = msg_id_str.parse::<u64>() {
                                                        let _ = http.delete_webhook_message(webhook_id, &config.webhook_token, Id::new(m_id)).await;
                                                    }
                                                }
                                                {
                                                    let conn = db.lock().unwrap();
                                                    let _ = crate::database::delete_purged_messages(&conn, &guild_id.get().to_string(), num);
                                                }
                                            }
                                        }

                                        {
                                            let conn = db.lock().unwrap();
                                            let mut entry = crate::database::get_count_entry(&conn, &guild_id.get().to_string()).unwrap();
                                            entry.count = num;
                                            entry.last_counter = "".to_string();
                                            let _ = crate::database::set_count_entry(&conn, &entry);
                                        }

                                        let mut msg = format!("Count set to {}.", num);
                                        if purge {
                                            msg.push_str(" Purged newer messages.");
                                        }

                                        Some(InteractionResponseData {
                                            content: Some(msg),
                                            flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                            ..Default::default()
                                        })
                                    }
                                    "channel" => {
                                        let mut channel_id = None;
                                        for o in opts {
                                            if o.name == "channel" {
                                                if let CommandOptionValue::Channel(c) = &o.value { channel_id = Some(*c); }
                                            }
                                        }
                                        if let Some(cid) = channel_id {
                                            if let Ok(wh_res) = http.create_webhook(cid, "Tally Counting").await {
                                                if let Ok(webhook) = wh_res.model().await {
                                                    let mut config = {
                                                        let conn = db.lock().unwrap();
                                                        crate::database::get_guild_config(&conn, &guild_id.get().to_string()).unwrap()
                                                    };
                                                    config.active = true;
                                                    config.channel = cid.get().to_string();
                                                    config.webhook_id = webhook.id.get().to_string();
                                                    config.webhook_token = webhook.token.unwrap_or_default();
                                                    {
                                                        let conn = db.lock().unwrap();
                                                        let _ = crate::database::set_guild_config(&conn, &config);
                                                    }
                                                    Some(InteractionResponseData {
                                                        content: Some(format!("Counting channel set to <#{}>", cid.get())),
                                                        flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                        ..Default::default()
                                                    })
                                                } else {
                                                    Some(InteractionResponseData {
                                                        content: Some("Failed to get webhook model.".to_string()),
                                                        flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                        ..Default::default()
                                                    })
                                                }
                                            } else {
                                                Some(InteractionResponseData {
                                                    content: Some("Failed to create webhook in channel.".to_string()),
                                                    flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                                    ..Default::default()
                                                })
                                            }
                                        } else {
                                            None
                                        }
                                    }
                                    "take-turns" => {
                                        let mut state = String::new();
                                        for o in opts {
                                            if o.name == "state" {
                                                if let CommandOptionValue::String(s) = &o.value { state = s.clone(); }
                                            }
                                        }
                                        let mut config = {
                                            let conn = db.lock().unwrap();
                                            crate::database::get_guild_config(&conn, &guild_id.get().to_string()).unwrap()
                                        };
                                        if state == "on" {
                                            config.allow_double_post = false;
                                        } else {
                                            config.allow_double_post = true;
                                        }
                                        {
                                            let conn = db.lock().unwrap();
                                            let _ = crate::database::set_guild_config(&conn, &config);
                                        }
                                        Some(InteractionResponseData {
                                            content: Some(format!("Take turns setting is now {}.", state)),
                                            flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                            ..Default::default()
                                        })
                                    }
                                    "view" => {
                                        let config = {
                                            let conn = db.lock().unwrap();
                                            crate::database::get_guild_config(&conn, &guild_id.get().to_string()).unwrap()
                                        };
                                        let channel_str = if config.channel.is_empty() {
                                            "None".to_string()
                                        } else {
                                            format!("<#{}>", config.channel)
                                        };
                                        let take_turns = if config.allow_double_post { "Off" } else { "On" };
                                        Some(InteractionResponseData {
                                            content: Some(format!("**Tally Settings:**\n\n**Counting Channel:** {}\n**Take Turns:** {}", channel_str, take_turns)),
                                            flags: Some(twilight_model::channel::message::MessageFlags::EPHEMERAL),
                                            ..Default::default()
                                        })
                                    }
                                    _ => None
                                }
                            }
                            _ => None
                        }
                    } else {
                        None
                    }
                } else {
                    None
                }
            }
            _ => None
        };

        if let Some(data) = response_data {
            let response = InteractionResponse {
                kind: InteractionResponseType::ChannelMessageWithSource,
                data: Some(data),
            };
            http.interaction(app_id)
                .create_response(interaction_id, &interaction_token, &response)
                .await?;
        }
    }
    
    Ok(())
}
